### ISSUE TITLE
Action doer definition

### ISSUE THREAD

**Issue by bmoneke** [pymeasure/experiment_control_protocol#17]
In #12 we discussed, how the ECP component doing actions should be named and what it does.
Especially the terms "actor" and "driver" and their relationship were not clear. 

Given:
- ECP messages.
- hardware devices
- actions on these devices or other things

Requirements:
- Language agnostic
- the Device or other software does what is told by the messages and sometimes acts on its own (regular readout etc.)

**Comment by bmoneke** [comment]
My original understanding (as a starting point):
We have a "message processor" which reads and writes ECP messages and maybe contains the regular readout part, a "driver" (which is a pymeasure instrument or similar).
Both parts together (and maybe some glue) is an actor.
The actor is a single program, which can be in any language, as it is accessed via ECP. It is, however, not possible to mix any language between processor and driver, as they have to work together closely.
That should be no problem, as the protocol is defined and the processor logic not difficult. Therefore an implementor can write a needed processor.

ECP gives one (or several) processors, such that it is easy to plug them together with drivers, as needed.

However, ecp mandates no specific interaction between processor and driver, but defines only the message protocol ("if you want to get a value, send `GET value.`"...) . The processors defined by ecp are only (working) examples how it could be done.

(i know, that it differs from your point @bilderbuchi, I just wanted to share a more complete picture of my original idea)

**Comment by bilderbuchi** [comment]
A couple of points:
> Given:
> "pymeasure instrument": a python class with properties and methods controlling a hardware device

This should imo read "some object controlling a hardware device" -- Pymeasure is just one of many choices, and we should not arrive at a solution specific to pymeasure.

> Requirements:
I would add "Agnostic of instrument control library" -- as above, we want a generically working solution.

> We have a "message processor" which reads and writes ECP messages and maybe contains the regular readout part, a "driver" (which is a pymeasure instrument or similar).
Both parts together (and maybe some glue) is an actor.

I'm a bit confused by this part.
* What does "regular" mean?
* What is the "readout part"?
* If a message processor _contains_ a driver, what is the distinction of that to "both parts together" (which is then called an actor)? Aren't "message processor" and "actor" effectively the same?

> It is, however, not possible to mix any language between processor and driver, as they have to work together closely.

I don't think we need that restriction. I don't see why we should disallow e.g. to have a Python processor and a C driver controlled via FFI or `ctypes`, for example.

Sorry that I'm a bit picky/precise on language -- I think we should take care to be precise, to avoid talking by each other.

**Comment by bilderbuchi** [comment]
I'm not sure if we "just" have a off-by-one/index shift in our names for things.
Let me try to sketch what I understand, with "stand-in names", and try to map that to our different proposals:
```mermaid
erDiagram
    ECP |o..|{ msg-processor : "exchanges messages with"
    action-doer ||--|| msg-processor : contains
    action-doer ||--|| hardware-interactor : contains
    msg-processor ||--|| hardware-interactor: "interacts with/controls"
```
|entity|Benedikt|Benjamin|Christoph|
|---|---|---|---|
|action-doer|Actor|Actor|Driver|
|msg-processor|message processor|not separately considered (action-doer exchanges messages)|not separately considered (action-doer exchanges messages)|
|hardware-interactor|Driver|Driver|not named|

Does that capture the situation correctly?

**Comment by bmoneke** [comment]
> This should imo read "some object controlling a hardware device" -- Pymeasure is just one of many choices, and we should not arrive at a solution specific to pymeasure.

I defined that name "pymeasure instrument" for us to have an actual example to talk about and use it in our discussion(it is a given, as pymeasure exists), not that it should enter the final definition of exp components.

>  > It is, however, not possible to mix any language between processor and driver, as they have to work together closely.
>
> I don't think we need that restriction. I don't see why we should disallow e.g. to have a Python processor and a C driver controlled via FFI or ctypes, for example.

With _any_ I mean really _any combination of two or more languages_. Some combinations are possible, but not necessarily any.

**Comment by bilderbuchi** [comment]
> With _any_ I mean really _any combination of two or more languages_. Some combinations are possible, but not necessarily any.

Hm, I understand "not any" to mean the same thing as "none" ("[keine](https://www.linguee.de/englisch-deutsch/uebersetzung/not+any.html)"), but then we're both not native speakers :sweat_smile: 
In any case, the edited formulation is better! I think that languages don't (shouldn't) enter the picture on the specification level, as we should be independent of that. If people want to write an action-doer in brainfuck, with a hardware-interactor in VBA macros in an Excel file, fine by us, as long as it conforms to the spec! :grin:.

**Comment by bilderbuchi** [comment]
Do you see the msg-processor as a generic part (e.g. a function or class) that does message processing for multiple Components, and implements the RPC part (mapping from message commands to e.g. function calls)?

**Comment by bmoneke** [comment]
What's that "process-logic-implementor" there? I thought the Director send commands (via cordinators) to the component we are about to define?

Do I understand it correctly, that the "action-doer" contains both the "msg processor" and the "hardware-interactor"? If so, the naming scheme is correct for me.

As a side note: I introduced (in above naming scheme) "message processor", as I did not know before, whether the "actor" is the whole thing or just the message part.

**Comment by bmoneke** [comment]
> Do you see the msg-processor as a generic part (e.g. a function or class) that does message processing for multiple Components, and implements the RPC part (mapping from message commands to e.g. function calls)?

Yes. On implementation side, the action-doer could be a subclass of the msg-processor, for example.

**Comment by bilderbuchi** [comment]
> What's that "process-logic-implementor" there? I thought the Director send commands (via cordinators) to the component we are about to define?

In my tool, I have found it useful to have some processing components that one can use arbitrarily, either pure/stateless or stateful. Some concrete examples are temperature conversion (K to C) and PID controllers. 
These are pulled in via config file, supplied input and output connections (to some Parameters), and do their thing based on an interval. The Director in my case (the GUI) triggers the updates of these things, but they fetch their data themselves.

In that view, these things (I called them "Processors") were Components just like Actors (in ECP lingo), and were useful to wire stuff together.
I included them here, because their naming should also be cleaned up while we do this.

> Do I understand it correctly, that the "action-doer" contains both the "msg processor" and the "hardware-interactor"? 

Yes

> As a side note: I introduced (in above naming scheme) "message processor", as I did not know before, whether the "actor" is the whole thing or just the message part.

Thanks, I was wondering about that. I was not sure why that was there. I did not think so far ahead concerning if we want to reuse a common RPC implementation.

In my view, the action-doer is the Component visible in the network, exchanging messages with it, and it _has a_ hardware-interactor that it uses to talk to a Device.
_How_ the action-doer maps received messages to function calls/attribute access in the hardware-interactor is something I consider an implementation detail, and as such outside the scope of the spec. This will be different depending on the language used.
As such, I would not have included the message-processor in the picture at all.

**Comment by bilderbuchi** [comment]
@bklebel @bmoneke what do you think about this proposal that
* follows (a mix of) your naming
* leaves out the msg-processor as an implementation detail
* description on a "glossary" detail level
```mermaid
erDiagram
    "ECP Component (Director, Coordinator, Processor,...)" |o..|{ Actor : "exchanges messages with"
    Actor ||--|| Driver: contains
    Driver||--|| "Device (HW)" : "communicates with"
```

> Actor
>     An Actor offers a standardized interface to the ECP network to communicate with some Device. This happens via a Driver contained in the Actor, see {ref}`components.md#actor`. An Actor implements the mapping/translation between ECP messages and the Driver's interface.
> 
> Device
>     Some piece of hardware controlled by a Driver.
> 
> Driver
>     A Driver is some object that takes care of communicating with a Device. This object is external to ECP, for example coming from and instrument control library like `pymeasure`, `instrumentkit` or `yaq`. See {ref}`components.md#driver`.
> 
> Processor
>     A Component on the ECP network which processes a number of inputs and produces a number of outputs. Can be stateful or stateless. See {ref}`components.md#processor`.

Is it ok for you to iterate on that until we are in agreement, then flesh out the details? Feedback, please! :-)

**Comment by bmoneke** [comment]
If we are talking about instrument control, I'm fine with your definitions.

A question regarding "Actor" and "Processor" remains: They are both network nodes, but the first one communicates with a Device while the latter provides some software defined features (PID controller etc.)?

Originally, I thought that "Actor" contains also the "Processor" category. I'm fine, with that (nominal) separation between going external ("Actor" acts on a Device) and staying inside the computer ("Processor" acts on software), even though both parts are quite similar.

**Comment by bilderbuchi** [comment]
Thanks for the feedback!

> If we are talking about instrument control, I'm fine with your definitions.

And if we are talking about something else...? What are you thinking of?

> A question regarding "Actor" and "Processor" remains: They are both network nodes, but the first one communicates with a Device while the latter provides some software defined features (PID controller etc.)?

Yes, correct. 
Using Processors, you could assemble a graph of computations/transformations. I could also see a Processor doing "sequences" (pymeasure "procedures"), and I could offer some ideas about that, but...later. :D

**Comment by bmoneke** [comment]
>What are you thinking of?

The "processor".


Should we mention, that the processor is similar to an actor, but implements software, which in turn can act, for example, transform data, do some orchestration like the director...?

**Comment by bilderbuchi** [comment]
Sure! I'll include that in the detailed description section for the processor.

**Comment by bilderbuchi** [comment]
I've updated #12 with the new nomenclature from https://github.com/pymeasure/experiment_control_protocol/issues/17#issuecomment-1399292666.
